[DPE-9311] Initial user documentation (#24)
canonical/valkey-operator: docs/how-to/tls.md

# How to enable TLS

Valkey provides a secure transport layer for communication between clients and Valkey,
and the charm provides a simple way of enabling TLS encryption.

> 

Reviewer: Can you add a section about the self-managed CA / certs? and on which layer it applies, and then introduce the alternative of using a TLS provider

Author: Added.

For internal communication between Valkey peers, Charmed Valkey always enables TLS
by default. The TLS certificates for this purpose are managed by the charm itself
and do not require a separate TLS provider. This applies to connections between
Valkey primary and replicas as well as between Valkey and Sentinel and between 
Sentinel instances.

To add another layer of security for communication between clients and Valkey,
follow the instructions in the next sections. The self-managed, internal peer TLS
will be replaced with an external TLS provider and encryption will be enabled 
for both client and peer TLS.

## Deploy a TLS provider

This guide will use the [Self-signed Certificates](https://charmhub.io/self-signed-certificates)
charm as an example for all cases.

```{caution}
**[Self-signed certificates](https://en.wikipedia.org/wiki/Self-signed_certificate) are not recommended for a production environment.**

Check the [Security with X.509 certificates](https://charmhub.io/topics/security-with-x-509-certificates) page for an overview of all the TLS certificates charms available. 
```

Deploy the `self-signed-certificates` charm. Charmed Valkey uses `v4` of the

> [{spellexception}`tls-certificates` library](https://charmhub.io/tls-certificates-interface/libraries/tls_certificates).

Reviewer 2: Question: How important is that information about the library version? It seems to be out of place here, in a How-to guide. We don't seem to use this information for anything here.

```suggestion
Deploy the `self-signed-certificates` charm:
```

Author: I think it can be very important for users, given the context that quite a few Data products only support earlier versions of the lib. This can lead to issues on deployment if Valkey is added to an existing deployment alongside another Data product such as MySQL, if not taken into consideration.

Reviewer 2: Oh, I see, thank you. I'd make it a warning then and maybe add some short practical advice on how that information can be used. Just like you explained it here in this comment. An explanation is fine in a how-to guide if it's a part of a warning and short.

Author: updated

```text
Running operation 1 with 1 task
  - task 2 on unit-self-signed-certificates-0

Waiting for task 2...
ca-certificate: |-
  -----BEGIN CERTIFICATE-----
  ...
  -----END CERTIFICATE-----
```

Now that you have the CA certificate, you can use it to verify the server certificate. 

Remember that the server is configured to also require client TLS. You need to provide
the client certificate and key to authenticate the client. 

For this guide, download the server's certificate and key from a unit of Charmed
Valkey to the local machine.

Download the certificate:

```shell
juju scp --container valkey valkey/1:/var/lib/valkey/tls/client.pem client.pem
```

Download the private key:

```shell
juju scp --container valkey valkey/1:/var/lib/valkey/tls/client.key client.key 
```

Now you have all required files and can provide them as options to the `valkey-cli` command:

```shell
valkey-cli -h 10.1.44.126 -p 6380 --tls --cert client.pem --key client.key --cacert ca_cert.pem
```

You should see an authentication error as result since the connection was established,
but no credentials provided:

```shell
10.1.44.126:6380> ping

> (error) NOAUTH Authentication required.

Reviewer 2: Suggestion: This code block has two issues: CLI prompt and output combined with input. Let's either distil the code block to contain only input (and then a dedicated code block for output) or use a terminal block for proper visualisation and functionality (including copying the input).

Reviewer 2: I'm afraid just changing the syntax to `terminal` is not enough here for it to be rendered correctly. See the preview above this link (just scroll up a bit): https://canonical-charmed-valkey--24.com.readthedocs.build/en/24/how-to/tls/#certificate-expiration-and-rotation

To fix this, please use the correct syntax: see the last example in this [syntax reference](https://canonical-starter-pack.readthedocs-hosted.com/stable/reference/myst-syntax-reference/#terminal-output). I would also recommend adding the `:copy:` option to enable the copy button.

Author: It's a bit difficult with the valkey-cli shell. I decided to make it output-only terminal blocks now.

Reviewer 2: <img width="855" height="364" alt="Image" src="https://github.com/user-attachments/assets/4ee764f3-4400-4fad-88c8-7fcff75f7758" />
and
<img width="845" height="242" alt="Image" src="https://github.com/user-attachments/assets/d98320a5-24c7-41f1-a3e1-bfd9fb414ba3" />

It's not looking good when rendered.
The `output-only` option is for when you only have CLI output, but in many cases here, you have both input and output. If you insist on keeping the CLI prompt here, I'd revert it back to the code blocks with `text` syntax to avoid syntax highlighting. But we will need to split input and output here, making it six small code blocks with some introductory sentence each instead of just one.

I've tried my best experimenting with the terminal block, but so far, I have not been able to reproduce the line `10.1.44.126:6380> ping` exactly in the input.

```

> You can now successfully connect to the server using the `tls` directive and provide

Reviewer 2: Question: We've demonstrated the negative scenario above. Can we add a demonstration of the positive scenario here? An example of using the `tls` directive with some sample output.

Author: I've added a few steps to make it a positive one.

This log message can be used for [creating an alert with COS](https://charmhub.io/topics/canonical-observability-stack/how-to/add-alert-rules).

As soon as new certificates are issued by the TLS provider, Valkey will replace 
the expiring certificate with the renewed one on each unit, reloading the TLS files
into Valkey to ensure continued communication.

## Manage private keys

You can manage private keys used by the charm to generate the certificate signing
requests (CSR) by storing the private key in a [juju secret](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/secret/)
and then referencing the secret in the [charm configuration](https://documentation.ubuntu.com/juju/latest/howto/manage-applications/index.html#configure-an-application).

### Store the private key in a Juju secret

To store the private key in a juju secret, run the following command:

```shell
juju add-secret tls-private-key private-key=$(base64 -w0 private-key.pem)
```

You can use the secret ID from the output to reference the secret in the charm configuration.

```shell

> secret:d6s4hbnmp25c765uceo0

Reviewer 2: praise: Good job at splitting output from input here as separate code blocks with dedicated descriptions.

tutorial  k8s-controller  microk8s/localhost  3.6.14   unsupported  19:06:13+01:00

App                       Version  Status  Scale  Charm                     Channel   Rev  Address         Exposed  Message
self-signed-certificates           active      1  self-signed-certificates  1/stable  586  10.152.183.111  no       
valkey                             active      3  valkey                    9/edge     11  10.152.183.123  no       

Unit                         Workload  Agent  Address      Ports  Message
self-signed-certificates/0*  active    idle   10.1.44.89          
valkey/0*                    active    idle   10.1.44.126         
valkey/1                     active    idle   10.1.44.117         
valkey/2                     active    idle   10.1.44.127         

Integration provider                   Requirer                    Interface         Type     Message
self-signed-certificates:certificates  valkey:client-certificates  tls-certificates  regular  
valkey:status-peers                    valkey:status-peers         status_peers      peer     
valkey:valkey-peers                    valkey:valkey-peers         valkey_peers      peer     

> ```

Reviewer 2: Suggestion: let's remove this part as redundant.

```suggestion
the TLS provider on the client-certificates endpoint:
```

> To disable the client-to-server communication, run:

Reviewer 2: Suggestion: Wording improvement for clarity.

```suggestion
To disable the client-to-server encryption, run:
```

Reviewer 2: Suggestion: If we remove the above code block, we no longer need this sentence:

```suggestion
```

Author: I've removed it.

self-signed-certificates           active      1  self-signed-certificates  1/stable  586  10.152.183.111  no       
valkey                             active      3  valkey                    9/edge     11  10.152.183.123  no       

Unit                         Workload  Agent  Address      Ports  Message
self-signed-certificates/0*  active    idle   10.1.44.89          
valkey/0*                    active    idle   10.1.44.126         
valkey/1                     active    idle   10.1.44.117         
valkey/2                     active    idle   10.1.44.127         

Integration provider  Requirer             Interface     Type  Message
valkey:status-peers   valkey:status-peers  status_peers  peer  
valkey:valkey-peers   valkey:valkey-peers  valkey_peers  peer  
```

You have successfully disabled encryption with TLS for Valkey. You can verify that
the database is running without encryption by checking the `valkey-cli` command
without the `tls` directive:

```shell
valkey-cli -h 10.1.44.126 -p 6379
```

You should see an authentication error as the result since the network connection was established,
but no credentials provided:

```shell

> 10.1.44.126:6379> ping

Reviewer 2: Suggestion: As above, let's fix the input/output combination as well as CLI prompt here.

(error) NOAUTH Authentication required.
```

Notice that the database is running without encryption for client connections only.

> For internal peer-to-peer communication, Charmed Valkey always uses TLS by default.

Reviewer 2: Question: Do we need the "by default" here?

```suggestion
For internal peer-to-peer communication, Charmed Valkey always uses TLS encryption.
```
